lib/email/sender.rb

```diff
@@ -514,5 +474,144 @@ def self.bounceable_reply_address?
+        @message.header['References']  = [topic_canonical_reference_id]
+      else
+        @message.header['Message-ID']  = Email::MessageIdService.generate_for_post(post)
+        @message.header['In-Reply-To'] = referenced_post_message_ids[0] || topic_canonical_reference_id
+        @message.header['References']  = [topic_canonical_reference_id, referenced_post_message_ids].flatten.compact.uniq
+      end
+    end
+
+    ##
+    # When sending an email for the first post (OP) of the topic, we do not
+    # set References or In-Reply-To headers, since there is nothing yet
+    # to reference. This counts as the first email in the thread.
+    #
+    # Once set, the post's `outbound_message_id` should _always_ be used
+    # when sending emails relating to a particular post to maintain threading.
+    # This will either be:
+    #
+    # a) A Message-ID generated in an external main client or service which
+    #    is recorded when creating a post from an IncomingEmail via Email::Receiver
+    # b) A Message-ID generated by Discourse and recorded when sending an email
+    #    for a newly created post, which is created and saved here to the
+    #    outbound_message_id column on the Post, to survive moves of that
+    #    posts between topics.
+    #
+    # The RFC that covers using "Identification Fields", which are References,
+    # In-Reply-To, Message-ID, et. al. can be found here. It's a good idea to read
+    # this beginning in the area immediately after these quotes, at least to understand
+    # the 3 main headers:
+    #
+    # > The "Message-ID:" field provides a unique message identifier that
+    # > refers to a particular version of a particular message.  The
+    # > uniqueness of the message identifier is guaranteed by the host that
+    # > generates it.
+    #
+    # > ...
+    #
+    # > The "In-Reply-To:" field may be used to identify the message (or
+    # > messages) to which the new message is a reply, while the "References:"
+    # > field may be used to identify a "thread" of conversation.
+    #
+    # https://www.rfc-editor.org/rfc/rfc5322.html#section-3.6.4
+    #
+    # It is a long read, but to understand the decision making process here
+    # you can take a look at:
+    #
+    # https://meta.discourse.org/t/discourse-email-messages-are-incorrectly-threaded/233499
+    def add_experimental_identification_field_headers(topic, post)
+      @message.header["Message-ID"] = Email::MessageIdService.generate_or_use_existing(post)
+
+      if post.post_number > 1
+        ##
+        # Whenever we reply to a post directly _or_ quote a post, a PostReply
+        # record is made, with the reply_post_id referencing the newly created
+        # post, and the post_id referencing the post that was quoted or replied to.
+        referenced_posts = find_referenced_posts(post)
+
+        # TODO (martin) Is it correct to refer to the OP in this case???
```

Comment:
I think so. A new post in a topic still has the topic OP as its originating context, otherwise it would be a shiny new topic.

Reply:
Cool thanks, will leave that in.

Reply:
And remove the TODO of course :)